### PR TITLE
[ONNX] Create stand alone diagnostic rule on nearest match finding in dispatcher

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104267

Change the diagnostic call in nearest match finding from UnsupportedNodeAnalysis to its own guarding rule.

### CODE REVIEW

**Review by thiagocrepaldi — APPROVED** [review]
LGTM, but @BowenBao is the best reference for Diagnostics

File: torch/onnx/_internal/diagnostics/rules.yaml [hunk 1]
```diff
@@ -354,6 +354,35 @@
     deprecated: false
     tags: []
 
+- id: FXE0014
+  name: nearest-match-symbolic-function
+  short_description:
+    text: Can not find the perfect match symbolic function with op schema, but find the nearest match one.
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/diagnostics/rules.yaml:360]: Make sense.

File: torch/onnx/_internal/fx/onnxfunction_dispatcher.py [hunk 2]
```diff
@@ -146,22 +175,23 @@ def _find_the_perfect_or_nearest_match_onnxfunction(
...
             f"a nearest match is found. Please check the ONNX output carefully. \n",
         )
-        diagnostic = diagnostics.UnsupportedFxNodeDiagnostic(
-            diagnostics.rules.no_symbolic_function_for_call_function,
-            diagnostics.levels.WARNING,
-            f"Cannot find a perfect match of symbolic overload for {aten_name}, "
-            f"which should be registered under {node.target}. But a nearest match is found.",
-            unsupported_fx_node=node,
+
+        diagnostic = diagnostic_context.inflight_diagnostic()
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/fx/onnxfunction_dispatcher.py:186]: Yes, it then becomes living under onnxfunction_dispatcher.py

File: torch/onnx/_internal/fx/fx_onnx_interpreter.py [hunk 3]
```diff
@@ -410,8 +410,8 @@ def run_node(
         else:
             raise RuntimeError(f"Found node type not defined in torch.fx: {node.op}")
 
-    @diagnostics.diagnose_call(diagnostics.rules.atenlib_fx_to_onnx)
     @_beartype.beartype
+    @diagnostics.diagnose_call(diagnostics.rules.atenlib_fx_to_onnx)
```
Comment (titaiwangms) [inline torch/onnx/_internal/fx/fx_onnx_interpreter.py:414]: Bug fixed. diagnose_call shoud be on top of the target function.

File: torch/onnx/_internal/diagnostics/rules.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/diagnostics/rules.yaml]: Rename it to `find-opschema-matched-symbolic-function`, and also covers the scenario when perfect match is found (LEVEL=NOTE). Added instructions on WARNING cases.

File: torch/onnx/_internal/diagnostics/rules.yaml [hunk 4]
```diff
@@ -354,6 +354,35 @@
...
+  full_description:
+    text: "Can not find the perfect match symbolic function with op schema, but find the nearest match one."
+    markdown: |
+      The warning message indicates that the dispatcher was unable to find an exact match for the symbolic
+      function based on the provided operation schema and input data types. Instead, it found the closest
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/diagnostics/rules.yaml]: LMK if you still see such things in the new md.

File: torch/onnx/_internal/diagnostics/rules.yaml [hunk 5]
```diff
@@ -354,6 +354,42 @@
...
+    markdown: |
+      When an ATen/Custom operator is registered and needs to be dispatched to an OnnxFunction, the input
+      dtypes of the ATen/Custom operator are compared with the input dtypes of the OnnxFunction opschemas
+      to find a match. However, if a perfect/exact match is not found, the dispatcher will attempt to find
+      the nearest match with the highest number of input dtypes matching the OnnxFunction opschemas, while
+      issuing a warning.
+
+      There are two types of level that can be triggered in this rule:
+
+      1. NOTE: A perfect match is found, and no warning is issued.
```
Comment (BowenBao) [inline torch/onnx/_internal/diagnostics/rules.yaml:373]:
```suggestion
      1. NONE: A perfect match is found, and no warning is issued.
```
Should be `NONE`. I hate sarif level that NOTE actually also means problem, just not as severe as warning.

Comment (BowenBao) [reply torch/onnx/_internal/diagnostics/rules.yaml:373]: Actually don't worry about it here. I found out Sarif extension render it as `Level: note, Kind informational`, although the json file lists it as `Level: none, Kind informational`. We should just drop NONE level from our diagnostics API. And point NOTE to sarif None ...... 💀

**Review by BowenBao — APPROVED** [review]

File: torch/onnx/_internal/diagnostics/rules.yaml (hunk 5 above)
Comment (titaiwangms) [reply torch/onnx/_internal/diagnostics/rules.yaml:373]: So keep it for now, and there will be a refactor for NOTE/NONE?

Comment (BowenBao) [reply torch/onnx/_internal/diagnostics/rules.yaml:373]: yep 🚢

**Pull request closed without merging**